src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+                BoundTypePattern => default,
+                BoundDeclarationPattern => default,
+                BoundConstantPattern => default,
+                BoundDiscardPattern => default,
+                BoundRelationalPattern => default,
+                null => default,
+                _ => throw ExceptionUtilities.UnexpectedValue(pattern)
+            };
+
+            static SetsOfOrCases rewriteBinary(BoundBinaryPattern binaryPattern)
+            {
+                SetsOfOrCases result = default;
+
+                if (binaryPattern.Disjunction)
+                {
+                    var patterns = ArrayBuilder<BoundPattern>.GetInstance();
+                    addPatternsFromOrTree(binaryPattern, patterns);
+
+                    // In `A1 or ... or An`, we produce an expansion set: `A1`, ..., `An`
+                    OrCases resultOrSet1 = result.StartNewOrCases();
+                    foreach (var pattern in patterns)
+                    {
+                        resultOrSet1.Add(pattern, pattern.Syntax);
+                    }
+
+                    // If any of the nested patterns can be expanded, we carry those on.
+                    // For example, in `... or Ai or ...`, when we found multiple `or` patterns in `Ai` to be expanded
+                    // For each such nested expansion, we'll produce an `... or <expansion> or ...` expansion
+                    for (int i = 0; i < patterns.Count; i++)
+                    {
+                        BoundPattern pattern = patterns[i];
+                        using SetsOfOrCases setOfOrCases = RewriteToSetsOfOrCases(pattern);
+                        if (!setOfOrCases.IsDefault)
+                        {
+                            foreach (OrCases orSet in setOfOrCases.Set)
+                            {
+                                OrCases resultOrSet = result.StartNewOrCases();
+                                foreach ((BoundPattern resultPattern, SyntaxNode? syntax) in orSet.Cases)
+                                {
+                                    BoundBinaryPattern? resultBinaryPattern = updateBinaryOrTree(binaryPattern, i, resultPattern).updated;
+                                    Debug.Assert(resultBinaryPattern is not null);
+                                    resultOrSet.Add(resultBinaryPattern, syntax);
+                                }
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    using SetsOfOrCases leftSetOfOrCases = RewriteToSetsOfOrCases(binaryPattern.Left);
+                    if (!leftSetOfOrCases.IsDefault)
+                    {
+                        // In `A and B`, when found multiple `or` patterns in `A` to be expanded
+                        // For each such nested expansion, we'll produce an expansion dropping the `and B`
+                        foreach (OrCases expandedLeftSet in leftSetOfOrCases.Set)
```

Comment:
>foreach (OrCases expandedLeftSet in leftSetOfOrCases.Set) [](http://example.com/codeflow?start=24&length=57)

Is this code simply creating a copy of ```leftSetOfOrCases```? If we drop the `using` from ```using SetsOfOrCases leftSetOfOrCases``` could we simply do ```result = leftSetOfOrCases```?